File: python/ch12_agent/agent_ch12/core/commands.py
"""操作系统命令执行边界。"""

from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True, slots=True)
class CommandResult:
    """操作系统命令执行后的统一结果。

    这是什么：命令执行的不可变返回值对象
    Java 类比：record CommandResult(String output, int exitCode, boolean timedOut, boolean truncated)
    为什么需要：不直接返回 subprocess 对象，让核心层不依赖具体操作系统 API

    参数：
        output: stdout 和 stderr 合并后的文本
        exit_code: 进程退出码，通常 0 表示成功
        timed_out: 是否因为超过时间限制被终止
        truncated: 输出是否超过上限并被截断
    """

    output: str  # stdout 和 stderr 合并后的文本。
    exit_code: int  # 进程退出码，通常 0 表示成功。
    timed_out: bool  # 是否因为超过时间限制被终止。
    truncated: bool  # 输出是否超过上限并被截断。


class CommandRunner(Protocol):
    """命令执行接口；真实实现和测试 Fake 都遵守它。

    这是什么：操作系统命令执行的抽象接口
    Java 类比：interface CommandRunner { CommandResult run(String cmd, String cwd); }
    为什么需要：AgentRunner 只依赖这个接口，测试时可以传入不会真正启动 PowerShell 的 Fake
    """

    def run(self, command: str, cwd: str, timeout_ms: int | None = None) -> CommandResult:
        """在受控工作目录中执行一条命令，并返回与操作系统无关的结果对象。

        参数：
            command: 要执行的命令文本
            cwd: 命令的工作目录
            timeout_ms: 可选的超时时间（毫秒），None 表示使用默认值
        """

File: typescript/chapters/ch02/src/core/commands.ts
/**
 * 命令执行边界：核心工具不直接依赖 child_process。
 * CommandResult 把输出、退出码、超时和截断统一为可观察状态，
 * 测试可以用确定性 CommandRunner 替换真实 PowerShellRunner。
 */
// 命令适配器的标准结果：输出限制和超时均为可观察状态，而非隐式异常。
export interface CommandResult {
  // 合并 stdout/stderr 的受限输出，供工具回填给模型。
  readonly output: string;
  // 子进程退出码；非零状态由工具层分类为错误结果。
  readonly exitCode: number;
  // 命令是否因时间预算耗尽而被终止。
  readonly timedOut: boolean;
  // 合并输出是否超过上限，防止模型把部分结果当作完整结果。
  readonly truncated: boolean;
}

// 核心工具只依赖此边界，可在测试中替换为确定性命令运行器。
export interface CommandRunner {
  // 可替换的进程执行边界，timeoutMs 仅覆盖本次调用。
  run(command: string, cwd: string, timeoutMs?: number): Promise<CommandResult>;
}
